File: various/lambda-functions/00-archived/lichess/lichess.py
""" lichess """
import datetime
import json
import os
import time
import urllib.request
from sendmail import SendMail


def handler(event, context):
    # pylint: disable=W0612,W0613
    """ Lambda function handler.
        Arguments:
            event - event from AWS
            context - execution context
    """
    Lichess().execute()

    return {
        'message': 'OK'
    }


class Lichess():
    """ Lichess backbone. """

    def execute(self):
        """ Lichess runner. """
        token = os.environ["LICHESS_TOKEN"]
        jsondata = self.fetch_overview(token)

        if jsondata is not None:
            overview = json.loads(jsondata.decode('utf-8'))
            games = self.get_my_turn_games(overview)
            if games:
                html = self.games_to_html(games)
                title = self.title()
                page = self.head(title)
                page = f"{page}<html><body><p>My turn in game vs.:</p>"
                page = f"{page}<ul>{html}</ul></body></html>"
                SendMail().deliver(title, page)


    @staticmethod
    def fetch_overview(token):
        """ Fetch overview of ongoing games.
            Args:
    	        token: lichess token
        """
        try:
            url = 'https://lichess.org/api/account/playing'
            request = urllib.request.Request(
                url, data=None,
                headers={
                    'User-Agent': ('Mozilla/5.0 (X11; Ubuntu; Linux x86_64;'
                                   'rv:55.0) Gecko/20100101 Firefox/55.0'),
                    'Authorization': 'Bearer ' + token
                }
            )
            return urllib.request.urlopen(request).read()
        except urllib.error.URLError:
            return None


    @staticmethod
    def get_my_turn_games(overview):
        """ Get the games in which I am in turn.
            Args:
                overview: the oveview json file
        """
        games = []
        for game in overview["nowPlaying"]:
            if game["isMyTurn"]:
                games.append(game)

        return games


    @staticmethod
    def title():
        """ Get title of mail. """
        now = time.strftime('%Y-%m-%d %H:%M:%S', time.localtime())
        return f"Lichess {now}"


    @staticmethod
    def head(title):
        """ Get HTML head (of page).
            Arguments:
                title - title.
        """
        return f"""<!DOCTYPE html>
<html dir="ltr" lang="en">
<head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
    <meta name="viewport" content="width=device-width" />
    <title>{title}</title>
</head>"""


    @staticmethod
    def games_to_html(games):
        """ Create HTML from games.
            Args:
                games: games in which I am in turn
        """
        html = ""
        formatter = "<li><i>{}</i>: {} (expires {})</li>"
        now = datetime.datetime.now()
        linkformatter = "<a href=\"https://lichess.org/{}\">Go to game</a>"
        for game in games:
            opponent = game["opponent"]["username"]
            link = linkformatter.format(game["fullId"])
            expires = now + datetime.timedelta(seconds=game["secondsLeft"])
            expires = expires.replace(microsecond=0)
            line = formatter.format(opponent, link, expires)
            html = f"{html}{line}"

        return html


if __name__ == '__main__':
    Lichess().execute()

File: various/lambda-functions/GoFuncs/LichessMoves/lichess.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"strings"
	"time"
)

func lichess(token string, title string) string {
	games := getGames(token)
	if !needToSendEmail(games) {
		return ""
	}

	return buildHTML(games, title)
}

func getGames(token string) []LichessGame {
	bearer := "Bearer " + token
	req, err := http.NewRequest("GET", "https://lichess.org/api/account/playing", nil)
	if err != nil {
		log.Fatal("Http Request - new: ", err)
	}

	req.Header.Add("Authorization", bearer)
	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		log.Fatal("Http Request - read: ", err)
	}

	body, _ := ioutil.ReadAll(resp.Body)

	var response LichessResponse
	err = json.Unmarshal([]byte(body), &response)
	if err != nil {
		log.Fatal("JSON Decode: ", err)
	}

	return response.NowPlaying
}

func needToSendEmail(games []LichessGame) bool {
	if games == nil || len(games) == 0 {
		return false
	}

	waitingOnMe := 0
	for _, game := range games {
		if game.IsMyTurn {
			waitingOnMe++
		}
	}

	return waitingOnMe > 0
}

func buildHTML(games []LichessGame, title string) string {
	var sb strings.Builder
	var myTurn string
	var link string

	sb.WriteString(`<!DOCTYPE html>
<html dir="ltr" lang="en">
<head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
    <meta name="viewport" content="width=device-width" />
	<title>`)
	sb.WriteString(title)
	sb.WriteString(`</title>
</head>
<body>
<p>My turn in game vs.:</p>
<ul>`)

	for _, game := range games {
		url := fmt.Sprintf("https://lichess.org/%s?bsts=%d", game.FullID, time.Now().Unix())
		if game.IsMyTurn {
			myTurn = "<b>YES</b>"
			link = fmt.Sprintf(" (<a href=\"%s\">Link</a>)", url)
		} else {
			myTurn = "no"
			link = ""
		}

		output := fmt.Sprintf("<li>My turn in game vs. <i>" + game.Opponent.Username + "</i>: " + myTurn + link + "</li>")
		sb.WriteString(output)
	}

	sb.WriteString("</ul></body></html>")
	return sb.String()
}
